### ISSUE TITLE
Support for Node Orientation?

### ISSUE THREAD

**Issue by adamnovak** [jltsiren/gcsa2#7]
Does GCSA2 support the concept of orientation for nodes? For example, if you have a k=5 de Bruijn graph for something like `GATTACACCCTGTAATC`, a sequence that starts with a 7-base motif and ends with its reverse complement, you would expect the node representing the first 5-mer in the de Bruijn graph and that representing the last 5-mer to be opposite orientations of the same node. I would expect this sort of structure, where you come out of a node, go around a loop, and head back into the same node in the opposite direction, would occur with some frequency in de Bruijn graphs with low k. However, I'm not sure how to articulate it in a form that GCSA2 will understand. Do you know how? Or are graphs not allowed to have that topology?

The reason I need this is that I'm working on node orientation and cycle support in vg. I already have node orientation in, but when you add cycles you are able to create graphs where you can't re-spell the graph to get around visiting a node in multiple orientations along the same path.

**Comment by jltsiren** [comment]
GCSA operates with three different graphs encoding the same information:
- The **original graph** has nodes with **string labels**. Each **position** in the original graph is a pair _(id, offset)_, where _id_ is the identifier of the node and _offset_ is the offset in its label.
- The **input graph** is similar to a finite automaton. Each **node** has a **character label** and corresponds to a position in the original graph. At the moment, `Node` provides the functionality for mapping between nodes and positions.
- GCSA construction creates a **path graph** (or a transformed graph, a prefix-sorted graph, or whatever other names I have been using), which is a generalization of the de Bruijn graph. Its nodes correspond to **paths** in the input graph, and it has the same **path labels** of length at most _2^d · k_, where _d_ is the number of doubling steps. This graph is never really exposed to the user.

The construction starts with a set of **paths** of length _k_ in the original graph, which are converted into paths in the input graph. Each path consists of a **start position**, a **label**, and a set of **extension positions**. The idea is that if path _B_ starts from an extension position of path _A_, path _AB_ is a valid path in the original graph, and we want to include it in the index.

Supporting orientations would basically require redefining positions as triples _(id, orientation, offset)_.

**Comment by adamnovak** [comment]
Is that a practical project in your opinion? Or should vg just hand GCSA2 a graph in which forward and reverse-complement versions of things are represented by different nodes (maybe with even and odd IDs)?

Looking through the code, if one were to add support for orientation in positions, it seems like you would also need to add it in some other places. For example, it also seems like it might be necessary to include the reverse complements of node sequences as nodes, so that they can be referred to by their ranks.

Does the current index search functionality search both strands, or only the forward strand?

**Comment by jltsiren** [comment]
From GCSA's point of view, there is no difference between the two approaches. The input graph will be the same, regardless of whether the nodes of the original graph have orientations, or the different orientations are handled as separate nodes. The BWT-based approach is based on sorting all paths in the input graph in lexicographic order. A path and its reverse complement will usually be separate nodes in the path graph, and hence separate paths in the input graph.

The search functionality only considers the forward strand, because there is no advantage in handling reverse complements inside GCSA. A simple wrapper function using the high-level interface would be essentially the same as any internal implementation.

**Comment by adamnovak** [comment]
OK, I'll continue on with my approach to feed in the forward and reverse
strands as separate nodes, and see how that goes.

**Comment by jltsiren** [comment]
I implemented native support for node orientation. `node_type` has now 53 bits for the identifier, 1 bit for the orientation, and 10 bits for the offset. The string representation of forward positions is still _id:offset_, while the representation of reverse complement positions is _id:-offset_. If the forward offsets run from _0_ to _k_, the corresponding reverse complement offsets are from _-k_ to _-0_ in the same order (_0_ corresponds to _-k_ and _k_ corresponds to _-0_).

This doesn't really change anything. GCSA sees `node_type` as a single integer, and it doesn't assume that the labels of different orientations of the same node are related. The construction handles the source and the sink node as special cases, so using the reverse complement orientation with them is probably a bad idea.

Initially I used the high-order bit of the offset for node orientation, maintaining backward compatibility, but it turned out that 9 bits was not enough for the offset. As the node identifier has been shifted one bit higher, indexes built before today no longer work correctly with the current version. (This isn't a major issue, because there are going to be major changes to the index components in the future.)
